File: Namyata/SquaresOfSortedArray.py
import unittest
from typing import List
# Given an integer array nums sorted in non-decreasing order, return an array of the squares of each number sorted in non-decreasing order.



def sortedSquares(nums: List[int]) -> List[int]:
    a = [i * i for i in nums]
    return sorted(a)

class TestSortedSquares(unittest.TestCase):

    def test_sorted_squares(self):
        nums = [-4, -1, 0, 3, 10]
        result = sortedSquares(nums)
        self.assertEqual(result, [0, 1, 9, 16, 100])

    def test_sorted_squares_negative_numbers(self):
        nums = [-7, -3, -2, 0, 5]
        result = sortedSquares(nums)
        self.assertEqual(result, [0, 4, 9, 25, 49])

    def test_sorted_squares_positive_numbers(self):
        nums = [1, 2, 3, 4, 5]
        result = sortedSquares(nums)
        self.assertEqual(result, [1, 4, 9, 16, 25])

    def test_sorted_squares_zero(self):
        nums = [0]
        result = sortedSquares(nums)
        self.assertEqual(result, [0])

    def test_sorted_squares_empty_list(self):
        nums = []
        result = sortedSquares(nums)
        self.assertEqual(result, [])

if __name__ == '__main__':
    unittest.main()

# All test cases worked!

File: CTests/squares_of_sorted_array.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

// Function declaration
int* sortedSquares(int* nums, int numsSize, int* returnSize);

// Function to return array of squares of each number sorted in non-decreasing order
int* sortedSquares(int* nums, int numsSize, int* returnSize) {
    int* result = (int*)malloc(numsSize * sizeof(int));
    *returnSize = numsSize;

    int left = 0;
    int right = numsSize - 1;
    int index = numsSize - 1;

    while (left <= right) {
        int leftSquare = nums[left] * nums[left];
        int rightSquare = nums[right] * nums[right];

        if (leftSquare > rightSquare) {
            result[index--] = leftSquare;
            left++;
        } else {
            result[index--] = rightSquare;
            right--;
        }
    }

    return result;
}

// Test cases
void test_sorted_squares() {
    bool success = true;
    int nums[] = {-4, -1, 0, 3, 10};
    int expected[] = {0, 1, 9, 16, 100};
    int numsSize = sizeof(nums) / sizeof(nums[0]);

    int returnSize;
    int* result = sortedSquares(nums, numsSize, &returnSize);

    printf("Test sorted_squares: ");
    if (numsSize == returnSize ) {
        for (int i = 0; i < returnSize; i++) {
            if(result[i] != expected[i]){
                 success = false;
                 break;
            }
               


        }
    }
    if(success) {
        printf(" Test sorted squares passed");
    } else {
        printf(" Test sorted squares failed");
    }

    printf("\n");

    free(result);
}

void test_sorted_squares_negative_numbers() {
    bool success = true;
    int nums[] = {-7, -3, -2, 0, 5};
    int expected[] = {0, 4, 9, 25, 49};
    int numsSize = sizeof(nums) / sizeof(nums[0]);

    int returnSize;
    int* result = sortedSquares(nums, numsSize, &returnSize);


   
    if (numsSize == returnSize ) {
        for (int i = 0; i < returnSize; i++) {
            if(result[i] != expected[i]){
                 success = false;
                 break;
            }
               


        }
    }
    if(success) {
        printf(" Test sorted_squares_negative_numbers passed");
    } else {
        printf(" Test sorted_squares_negative_numbers failed");
    }
     printf("\n");
    free(result);
}

void test_sorted_squares_positive_numbers() {
    bool success = true;
    int nums[] = {1, 2, 3, 4, 5};
    int expected[] = {1, 4, 9, 16, 25};
    int numsSize = sizeof(nums) / sizeof(nums[0]);

    int returnSize;
    int* result = sortedSquares(nums, numsSize, &returnSize);


     if (numsSize == returnSize ) {
        for (int i = 0; i < returnSize; i++) {
            if(result[i] != expected[i]){
                 success = false;
                 break;
            }
               


        }
    }
    if(success) {
        printf(" Test sorted_squares_positive_numbers passed");
    } else {
        printf(" Test sorted_squares_positive_numbers failed");
    }
    printf("\n");

    free(result);
}




int main() {
    test_sorted_squares();
    test_sorted_squares_negative_numbers();
    test_sorted_squares_positive_numbers();


    return 0;
}
